### ISSUE TITLE
Cyrillic text issue after updating to haxe 4.0.0-rc.5 (in openfl TextField)

### ISSUE THREAD

**Issue by www0z0k** [openfl/openfl#2264]
After upgrading haxe I get correct traces of Cyrillic strings’ lengths and values, but the TextField itself looks like:
![Screenshot 2019-09-21 at 20 21 43](https://user-images.githubusercontent.com/951595/65383467-a942de80-dd1e-11e9-91dc-289a91c15c68.png)

However tracing the textField.text gives a correct result in the console.

This is on mac/ios targets. `toUpperCase()` and `length` work as expected for Cyrillic strings, but the text looks like:
![Screenshot 2019-09-22 at 03 15 47](https://user-images.githubusercontent.com/951595/65383464-9f20e000-dd1e-11e9-8640-d0b693e71ada.png)

On neko target `toUpperCase()` and `length` don't work correctly, but the text is displayed ok:
![Screenshot 2019-09-22 at 03 17 14](https://user-images.githubusercontent.com/951595/65383470-b8299100-dd1e-11e9-9e34-4de25a1f7331.png)

Here is the full code to reproduce it:
```
	var _body:TextField = new TextField();
        addChild(_body);
	_body.text = 'фыва йцук';
```
Should I downgrade back to 3.4.7 version or there might be another way around?

**Comment by MSGhero** [comment]
I need to look into it

**Comment by MSGhero** [comment]
@www0z0k what versions of OpenFL, Lime, and hxcpp are you on?

**Comment by www0z0k** [comment]
OpenFL Command-Line Tools (8.9.5-LoqcFr)
Lime Command-Line Tools (7.6.3)
how do I check hxcpp version?

**Comment by Txori** [comment]
Hi. I think I might have the same issue regarding Japanese characters. It works on Flash but not on Android: http://forum.haxeflixel.com/topic/934/display-japanese-font-on-android-target/

flixel: [4.6.3]
openfl: [8.9.4]
lime: [7.6.2]
hxcpp: [4.0.52]

@www0z0k , type "haxelib list" in cmd

**Comment by MSGhero** [comment]
I may not have time for a few days to look into this, but can one of you go here (https://github.com/openfl/openfl/blob/develop/src/openfl/_internal/text/UTF8String.hx) and change line 6 to `= UnicodeString` instead of normal `String`?

**Comment by www0z0k** [comment]
@MSGhero where can I find openfl internals locally?

**Comment by Txori** [comment]
@www0z0k 
HaxeToolKit/haxe/lib/openfl/8,9,4/src/openfl/_internal/text/UTF8String.hx

But I just tried and it's not working.

**Comment by www0z0k** [comment]
@Txori I wonder where is it located on osx
Are you sure that it takes the source files each time you run and nothing gets precompiled?

**Comment by Txori** [comment]
Sorry I know almost nothing about osx... Isn't it a bit like linkux? Maybe something like HD/usr/lib/haxe ?

In order to verify it takes modification of the source files, I suppressed .\export\android\obj\src\openfl\_internal\text folder and it gets regenerated.

**Comment by MSGhero** [comment]
I'm getting compiler errors from within hxcpp, so I can't even test this. The text looks fine on HTML5 target, and if you're saying it also looks fine on Neko, then this might actually be a hxcpp issue?

I'll try to reinstall haxe and stuff tomorrow and see if that helps. Otherwise, I'm only guessing at what the real problem is here.

**Comment by www0z0k** [comment]
@MSGhero 
please check the displayed text along with:
```
trace('йцу'.length);
trace('йцу'.charAt(0));
```
It shows `length` 6 for actual 3 letters and `charAt` `?` on platforms which display text correctly and shows correct traces on platforms that display incorrect text, which makes me assume it has something to do with the new full ubicode support in haxe 4

**Comment by MSGhero** [comment]
I cannot compile when my code has non-ASCII characters in it. Like I get compiler errors from hxcpp based on the contents of my strings. 

If you can try this on HTML5 and see what the lengths are, that would be helpful.

**Comment by www0z0k** [comment]
@MSGhero this is not what I expected:
```
var _body:TextField = new TextField();
addChild(_body);
_body.width = 500;
_body.text = 'фыва, length = ' + 'фыва'.length + ' at 0 = "' + 'фыва'.charAt(0) + '"';
```
shows:
![Screenshot 2019-09-24 at 16 42 13](https://user-images.githubusercontent.com/951595/65516795-5453c380-deea-11e9-9e42-60b4300b9507.png)


Tested on osx if it matters

**Comment by Gama11** [comment]
@MSGhero I recall some hxcpp issue where that happens when your Visual Studio version / C++ compiler is too old.

**Comment by MSGhero** [comment]
Ohhhh interesting. Thanks @Gama11 I’ll try it out

**Comment by MSGhero** [comment]
@jgranick this issue and #2256 (probably) are caused by these lines on rc4 and rc5: https://github.com/openfl/openfl/blob/develop/src/openfl/_internal/text/TextLayout.hx#L143-L148

This is a Lime/Harfbuzz interface, so you probably have a much better understanding of it than I do. Non-UTF8 codepoints look like they're being wrapped or truncated.

**Comment by Txori** [comment]
When I revert back openfl/src/openfl/_internal/text/TextLayout.hx to [b151041](url=https://github.com/openfl/openfl/commit/b151041b497a7edd5aec3a49d394c742d8360668#diff-fecc2790a3850e043f8e6115c482989d), every thing is working perfectly. What caused the roll back?

**Comment by filipp8** [comment]
I have the same problem with japanese text on rc3

**Comment by Txori** [comment]
Just un-comment those lines in openfl/src/openfl/_internal/text/TextLayout.hx

`// #if ((haxe_ver < "4.0.0") || neko || mac || linux || hl)
__hbBuffer.addUTF8(text, 0, -1);
// #else
// for (i in 0...text.length)
// 	__hbBuffer.add(text.charCodeAt(i), i);
// #end`

It works, unless you try a Windows build: https://github.com/openfl/openfl/commit/8dcb6ba927c88926784e990a2a2520cae00a5130

**Comment by rainyt** [comment]
https://github.com/openfl/openfl/commit/8dcb6ba927c88926784e990a2a2520cae00a5130#commitcomment-35418723
Can see if this helps.

**Comment by MSGhero** [comment]
@rainyt uh yeah it works. That's way simpler than anything I was going to try. I want to run this through the unit tests first to be sure, but wow.

**Comment by jgranick** [comment]
Is there any way to set the type to unicode and then set the text in one call instead of character by character?

**Comment by MSGhero** [comment]
@jgranick I checked the HB source, and all of the addUTF functions set the content type to UNICODE. So it likely comes down to fitting `addUTF16`’s API and then making sure that’s appropriate for all character types.

**Comment by MSGhero** [comment]
I've been super busy, but how about this? 
`__hbBuffer.addUTF16(DataPointer.fromBytes(Bytes.ofString(text, RawNative)), text.length, 0, -1);`

It's basically what Joshua said before, but I didn't see a cleaner way of writing it

**Comment by www0z0k** [comment]
any chance of a fix in some nightly build or something?

**Comment by MSGhero** [comment]
I’m close, I just don’t have much free time for coding like I used to. Sorry about that. 

The proposed solution where the `add` goes into a for loop doesn’t account for a lot of other work the `addUTF` functions perform, so it’s not a long-term fix that can be committed.

**Comment by stnguyen** [comment]
anybody has a fix for it?

**Comment by ngrebenshikov** [comment]
My env: target iOS, lime (git), hxcpp (git), haxe 4.0.1
It's strange but the code @MSGhero mentioned (__hbBuffer.addUTF16(DataPointer.fromBytes(Bytes.ofString(text, RawNative)), text.length, 0, -1);) doesn't work for me.

But uncommenting previous fix helped me after setting the contentType to the unicode. So my code inside TextLayout.hx

```haxe
__hbBuffer.contentType = lime.text.harfbuzz.HBBufferContentType.UNICODE;
for (i in 0...text.length) {
    __hbBuffer.add(text.charCodeAt(i), i);
}
```

**Comment by stnguyen** [comment]
@ngrebenshikov @rainyt do you use haxe 4 `String`, or another string class?
Would you pls post a fully working example, with note about versions of haxe, openfl, lime, and platform?
Thank you so much!

**Comment by ngrebenshikov** [comment]
@stnguyen I can only share may environment. I don't have a small working example to share.
Haxe - 4.0.1
String - Haxe 4 String
Hxcpp - latest Github version
Lime - latest GitHub version
OpenFL - 8.9.5
Build platform - OSX Catalina
Target platform - iOS

**Comment by MSGhero** [comment]
Ok. I apologize for the long delay. I'm not very familiar with cpp, I've moved, I'm busy, etc, but I got it. Below is on the latest Haxe with a 2 line modification to TextLayout and DataPointer. Joshua pointed out before that I should test this on several C targets just to be sure nothing breaks across the board.

![image](https://user-images.githubusercontent.com/5033927/70020890-ae14e180-155c-11ea-8d5b-5eb0930689f1.png)
The mods are:
```haxe
// __hbBuffer.addUTF8(text, 0, -1);
__hbBuffer.addUTF16(untyped __cpp__('(uintptr_t){0}', text.wc_str()), text.length, 0, -1);
```
and giving `DataPointer` a `from DataPointerType` in the abstract declaration.

What I was doing in the past was getting the address of `text` which is a Haxe object which means nothing to the text shaping engine. `wc_str()` gives you the UTF16 pointer of the string, which means nothing to Haxe and specifically nothing to the `DataPointer` class. The above line is a direct cpp call that I found in `DataPointer` that has to be called from within `TextLayout` to make use of the UTF16 pointer, which then enables `DataPointer` to be used as intended.

PR soon.
